torch/_decomp/decompositions.py

```diff
@@ -3057,6 +3047,44 @@ def nll_loss_forward(
     return result, total_weight
 
 
+@register_decomposition(aten.nll_loss_forward)
+def nll_loss_forward(
+    self: Tensor,
+    target: Tensor,
+    weight: Optional[Tensor],
+    reduction: int,
+    ignore_index: int,
+) -> Tuple[Tensor, Tensor]:
+    assert self.dim() > 0 and self.dim() <= 2, "input tensor should be 1D or 2D"
+    assert (
+        target.dim() <= 1
+    ), "0D or 1D target tensor expected, multi-target not supported"
+
+    no_batch_dim = self.dim() == 1 and target.dim() == 0
+    assert no_batch_dim or (
+        self.shape[0] == target.shape[0]
+    ), f"size mismatch (got input: {self.shape}, target: {target.shape})"
+
+    n_classes = self.shape[-1]
+
+    assert weight is None or (
+        weight.dim() == 1 and weight.numel() == n_classes
+    ), f"weight tensor should be defined either for all {n_classes} classes or no classes but got weight tensor of shape: {weight.shape}"  # noqa: B950
+
+    return _nll_loss_forward(self, target, weight, reduction, ignore_index)
+
+
+@register_decomposition(aten.nll_loss2d_forward)
+def nll_loss2d_forward(
+    self: Tensor,
+    target: Tensor,
+    weight: Optional[Tensor],
+    reduction: int,
+    ignore_index: int,
+) -> Tuple[Tensor, Tensor]:
+    return _nll_loss_forward(self, target, weight, reduction, ignore_index)
```

Comment:
I'm kind of confused about this operator. It doesn't seem to be used anywhere in Python. Do we actually hit this operator?

Reply:
The reason I ask is that I would expect some nll_loss2d_forward-specific shape checks here.

Reply:
`nll_loss_nd` (CompositeImplicit) is the entry-point for this loss  (`torch.nn.functional.nll_loss` calls this as well as `cross_entropy_loss`).

`nll_loss_nd` has error checking logic and calls into `nll_loss` or `nll_loss_2d` (which again are CompositeImplicit)
https://github.com/pytorch/pytorch/blob/a8e3bd97cfd94b92fc202f330df6bc2f808846a6/aten/src/ATen/native/LossNLL.cpp#L679-L685

Both `nll_loss` and `nll_loss_2d` call into the operators `nll_loss_forward` and `nll_loss2d_forward` (which are CompositeExplicit).

https://github.com/pytorch/pytorch/blob/a8e3bd97cfd94b92fc202f330df6bc2f808846a6/aten/src/ATen/native/LossNLL.cpp#L662-L668
https://github.com/pytorch/pytorch/blob/a8e3bd97cfd94b92fc202f330df6bc2f808846a6/aten/src/ATen/native/LossNLL2d.cpp#L489-L495

Also, our batching rules were registered for these `nll_loss_forward` and `nll_loss2d_forward` https://github.com/pytorch/pytorch/blob/a8e3bd97cfd94b92fc202f330df6bc2f808846a6/aten/src/ATen/functorch/BatchRulesLoss.cpp#L310-L313

Another thing to note is that there is a decomposition for `nll_loss` which is CompositeImplicit, have filed an issue for the same https://github.com/pytorch/pytorch/issues/106375